File: src/main/java/org/vors/pairbot/model/UserInfo.java
package org.vors.pairbot.model;

import org.hibernate.annotations.CacheConcurrencyStrategy;
import org.springframework.data.annotation.CreatedDate;

import javax.persistence.*;
import java.util.Date;
import java.util.HashSet;
import java.util.Set;

@Entity
@Cacheable
@org.hibernate.annotations.Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
@Table(name = "users")
public class UserInfo {

    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Id
    private Long pk;

    @Column
    private Integer userId;

    @Column
    private String firstName;

    @Column
    private String lastName;

    @OneToMany(fetch = FetchType.LAZY,
            mappedBy = "user",
            cascade = CascadeType.ALL,
            orphanRemoval = true)
    private Set<UserSetting> settings = new HashSet<>();

    @Column
    @Temporal(TemporalType.TIMESTAMP)
    @CreatedDate
    private Date createdDate;

    @Column
    @Temporal(TemporalType.TIMESTAMP)
    private Date lastPairDate;

    @ManyToOne
    @JoinColumn(name = "team_pk")
    private Team team;

    public Integer getUserId() {
        return userId;
    }

    public void setUserId(Integer userId) {
        this.userId = userId;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    public Set<UserSetting> getSettings() {
        return settings;
    }

    public void setSettings(Set<UserSetting> settings) {
        this.settings = settings;
    }

    public Date getCreatedDate() {
        return createdDate;
    }

    public void setCreatedDate(Date createdDate) {
        this.createdDate = createdDate;
    }

    public Long getPk() {
        return pk;
    }

    public void setPk(Long pk) {
        this.pk = pk;
    }

    public Team getTeam() {
        return team;
    }

    public void setTeam(Team team) {
        this.team = team;
    }

    public Date getLastPairDate() {
        return lastPairDate;
    }

    public void setLastPairDate(Date lastPairDate) {
        this.lastPairDate = lastPairDate;
    }
}

File: src/main/java/org/vors/pairbot/model/UserInfo.kt
package org.vors.pairbot.model

import org.hibernate.annotations.CacheConcurrencyStrategy
import org.springframework.data.annotation.CreatedDate
import java.time.ZoneId
import java.util.*
import javax.persistence.*

@Entity
@Cacheable
@org.hibernate.annotations.Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
@Table(name = "users")
class UserInfo(
        var userId: Long,
        var firstName: String,

        var lastName: String? = null,

        @Temporal(TemporalType.TIMESTAMP)
        var lastDeclineDate: Date? = null,
        var lastMessageId: Int? = null,
        var timezone: ZoneId? = null,
        @ManyToOne @JoinColumn(name = "team_pk", referencedColumnName = "pk")
        var team: Team? = null,

        @Temporal(TemporalType.TIMESTAMP) @CreatedDate
        var createdDate: Date? = null,

        @OneToMany(fetch = FetchType.LAZY, mappedBy = "user", cascade = [CascadeType.ALL], orphanRemoval = true)
        var settings: MutableSet<UserSetting> = HashSet(),
        var xp: Int = 0,
        @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
        var pk: Long = 0,
        @ManyToOne @JoinColumn(name = "last_partner_pk", referencedColumnName = "pk")
        var lastPartner: UserInfo? = null
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as UserInfo

        if (pk != other.pk) return false

        return true
    }

    override fun hashCode(): Int {
        return pk.hashCode()
    }

    override fun toString(): String {
        return "pk $pk"
    }
}
